File: app/src/main/java/org/kl/firearrow/fs/FileManager.java
/*
 * Licensed under the MIT License <http://opensource.org/licenses/MIT>.
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2022-2025 https://github.com/klappdev
 *
 * Permission is hereby  granted, free of charge, to any  person obtaining a copy
 * of this software and associated  documentation files (the "Software"), to deal
 * in the Software  without restriction, including without  limitation the rights
 * to  use, copy,  modify, merge,  publish, distribute,  sublicense, and/or  sell
 * copies  of  the Software,  and  to  permit persons  to  whom  the Software  is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE  IS PROVIDED "AS  IS", WITHOUT WARRANTY  OF ANY KIND,  EXPRESS OR
 * IMPLIED,  INCLUDING BUT  NOT  LIMITED TO  THE  WARRANTIES OF  MERCHANTABILITY,
 * FITNESS FOR  A PARTICULAR PURPOSE AND  NONINFRINGEMENT. IN NO EVENT  SHALL THE
 * AUTHORS  OR COPYRIGHT  HOLDERS  BE  LIABLE FOR  ANY  CLAIM,  DAMAGES OR  OTHER
 * LIABILITY, WHETHER IN AN ACTION OF  CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package org.kl.firearrow.fs;

import android.content.Context;
import androidx.annotation.NonNull;

import java.io.File;
import java.io.FileWriter;
import java.lang.ref.Cleaner;

import org.kl.firearrow.time.StopWatch;

public final class FileManager implements AutoCloseable {

    private final static String ERASE_DIRECTORY = "erase";
    private final static String ERASE_FILE1 = "erase1.txt";
    private final static String ERASE_FILE2 = "erase2.txt";
    private final static int COUNT_LINES = 50;

    private long nativeHandle;

    private static final Cleaner cleaner = Cleaner.create();
    private final Cleaner.Cleanable cleanable;
    private final Destructor destructor;

    private final StopWatch stopWatch;

    public FileManager() {
        if (nativeHandle == 0) {
            nativeHandle = nativeCreate();
        }

        this.destructor = new Destructor(nativeHandle);
        this.cleanable = cleaner.register(this, destructor);

        this.stopWatch = new StopWatch();
    }

    private static class Destructor implements Runnable {
        private long nativeHandle;

        private Destructor(long nativeHandle) {
            this.nativeHandle = nativeHandle;
        }

        @Override
        public void run() {
            if (nativeHandle != 0) {
                nativeDestroy(nativeHandle);
                nativeHandle = 0;
            }
        }
    }

    @Override
    public void close() {
        cleanable.clean();
    }

    private static native long nativeCreate();

    private static native void nativeDestroy(long nativeHandle);

    private static native long nativeEraseFile(long nativeHandle, @NonNull String path) throws FileException;

    private static native long nativeEraseFile(long nativeHandle, @NonNull String path, OverwriteMode mode) throws FileException;

    private static native long nativeEraseDirectory(long nativeHandle, @NonNull String path, boolean recursive) throws FileException;

    private static native long nativeEraseDirectory(long nativeHandle, @NonNull String path, OverwriteMode mode, boolean recursive) throws FileException;

    public long eraseFile(@NonNull String path) throws FileException {
        return nativeEraseFile(nativeHandle, path);
    }

    public long eraseFile(@NonNull String path, OverwriteMode mode) throws FileException {
        return nativeEraseFile(nativeHandle, path, mode);
    }

    public long eraseDirectory(@NonNull String path, boolean recursive) throws FileException {
        return nativeEraseDirectory(nativeHandle, path, recursive);
    }

    public long eraseDirectory(@NonNull String path, OverwriteMode mode, boolean recursive) throws FileException {
        return nativeEraseDirectory(nativeHandle, path, mode, recursive);
    }

    public String javaDeleteFile(@NonNull Context context) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart Java deleting file\n");

        final var directory = createDirectory(context, builder);
        final var file = new File(directory, ERASE_FILE2);
        final var filePath = file.getPath();
        fillFile(file, "Some text", builder);

        builder.append("> Delete file ").append(filePath).append("\n");

        if (!file.delete()) {
            builder.append("> File ").append(filePath).append(" didn't deleted\n");
        }

        stopWatch.stop();
        builder.append("> Java execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }

    public String cppEraseFile(@NonNull Context context, OverwriteMode mode) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart C++ erasing file with ").append(mode.name()).append("\n");

        final var directory = createDirectory(context, builder);
        final var file = new File(directory, ERASE_FILE1);
        final var filePath = file.getPath();
        fillFile(file, "Some text", builder);

        try {
            builder.append("> Erase file ").append(filePath).append("\n");

            final long duration = eraseFile(filePath, mode);

            builder.append("> C++ execution time: ").append(duration).append(" ms\n");
        } catch (FileException e) {
            builder.append("> Erase file ").append(filePath).append(" exception")
                   .append(e.getMessage()).append("\n");
        }

        stopWatch.stop();
        builder.append("> JNI execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }

    public String javaDeleteDirectory(@NonNull Context context) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart Java deleting directory\n");

        final var directory = createDirectory(context, builder);
        final String directoryPath = directory.getPath();

        final var file1 = new File(directory, ERASE_FILE1);
        fillFile(file1, "Some text in file1", builder);

        final var file2 = new File(directory, ERASE_FILE2);
        fillFile(file2, "Some text in file2", builder);

        builder.append("> Delete directory ").append(directoryPath).append("\n");
        deleteDirectory(directory, builder);

        stopWatch.stop();
        builder.append("> Java execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }

    public String cppEraseDirectory(@NonNull Context context) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart C++ erasing directory\n");

        final var directory = createDirectory(context, builder);
        final String directoryPath = directory.getPath();

        final var file1 = new File(directory, ERASE_FILE1);
        fillFile(file1, "Some text in file1", builder);

        final var file2 = new File(directory, ERASE_FILE2);
        fillFile(file2, "Some text in file2", builder);

        try {
            builder.append("> Erase directory ").append(directoryPath).append("\n");

            final long duration = eraseDirectory(directoryPath, false);

            builder.append("> C++ execution time: ").append(duration).append(" ms\n");
        } catch (FileException e) {
            builder.append("> Erase directory ").append(directoryPath).append(" exception")
                   .append(e.getMessage()).append("\n");
        }

        stopWatch.stop();
        builder.append("> JNI execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }

    public String javaEraseDirectoryRecursive(@NonNull Context context) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart Java deleting directory recursive\n");

        final var directory = createDirectory(context, builder);
        final String directoryPath = directory.getPath();

        final var file1 = new File(directory, ERASE_FILE1);
        fillFile(file1, "Some text in file1", builder);

        final var file2 = new File(directory, ERASE_FILE2);
        fillFile(file2, "Some text in file2", builder);

        builder.append("> Delete directory recursive").append(directoryPath).append("\n");
        deleteDirectoryRecursive(directory, builder);

        stopWatch.stop();
        builder.append("> Java execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }

    public String cppEraseDirectoryRecursive(@NonNull Context context) {
        stopWatch.reset();
        stopWatch.start();

        final var builder = new StringBuilder();
        builder.append("\nStart C++ erasing directory recursive\n");

        final var directory = createDirectory(context, builder);
        final String directoryPath = directory.getPath();

        final var file1 = new File(directory, ERASE_FILE1);
        fillFile(file1, "Some text in file1", builder);

        final var file2 = new File(directory, ERASE_FILE2);
        fillFile(file2, "Some text in file2", builder);

        try {
            builder.append("> Erase directory recursive").append(directoryPath).append("\n");

            final long duration = eraseDirectory(directoryPath, true);

            builder.append("> C++ execution time: ").append(duration).append(" ms\n");
        } catch (FileException e) {
            builder.append("> Erase directory recursive").append(directoryPath)
                   .append(" exception").append(e.getMessage()).append("\n");
        }

        stopWatch.stop();
        builder.append("> JNI execution time: ").append(stopWatch.getDuration()).append(" ms\n");

        return builder.toString();
    }


    private File createDirectory(@NonNull Context context, @NonNull StringBuilder builder) {
        final var directory = new File(context.getFilesDir(), ERASE_DIRECTORY);

        if (!directory.exists()) {
            if (!directory.mkdir()) {
                builder.append("> Directory ").append(ERASE_DIRECTORY).append(" didn't created\n");
            }
        }

        return directory;
    }

    private void fillFile(@NonNull File file, @NonNull String text,
                          @NonNull StringBuilder builder) {
        try {
            final var writer = new FileWriter(file);

            for (int i = 0; i < COUNT_LINES; i++) {
                writer.append(text).append(" ")
                      .append(String.valueOf(i)).append("\n");
            }

            writer.flush();
            writer.close();
        } catch (Exception e) {
            builder.append("> File write exception: ").append(e.getMessage()).append("\n");
        }
    }

    private void deleteDirectory(@NonNull File directory, @NonNull StringBuilder builder) {
        if (!directory.isDirectory()) {
            return;
        }

        final String[] children = directory.list();

        if (children == null) {
            builder.append("> Directory didn't contain files\n");
            return;
        }

        for (String child : children) {
            final var childFile = new File(directory, child);

            if (!childFile.delete()) {
                builder.append("> File ").append(childFile.getPath()).append(" didn't deleted\n");
            }
        }

        if (!directory.delete()) {
            builder.append("> Directory ").append(directory.getPath()).append(" didn't deleted\n");
        }
    }

    private void deleteDirectoryRecursive(@NonNull File fileOrDirectory, @NonNull StringBuilder builder) {
        final File[] children = fileOrDirectory.listFiles();

        if (children != null) {
            for (File child : children) {
                deleteDirectoryRecursive(child, builder);
            }
        }

        if (!fileOrDirectory.delete()) {
            builder.append("> Directory ").append(fileOrDirectory.getPath()).append(" didn't deleted\n");
        }
    }
}

File: app/src/main/cpp/fs/FileManager.cpp
/*
 * Licensed under the MIT License <http://opensource.org/licenses/MIT>.
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2022-2025 https://github.com/klappdev
 *
 * Permission is hereby  granted, free of charge, to any  person obtaining a copy
 * of this software and associated  documentation files (the "Software"), to deal
 * in the Software  without restriction, including without  limitation the rights
 * to  use, copy,  modify, merge,  publish, distribute,  sublicense, and/or  sell
 * copies  of  the Software,  and  to  permit persons  to  whom  the Software  is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE  IS PROVIDED "AS  IS", WITHOUT WARRANTY  OF ANY KIND,  EXPRESS OR
 * IMPLIED,  INCLUDING BUT  NOT  LIMITED TO  THE  WARRANTIES OF  MERCHANTABILITY,
 * FITNESS FOR  A PARTICULAR PURPOSE AND  NONINFRINGEMENT. IN NO EVENT  SHALL THE
 * AUTHORS  OR COPYRIGHT  HOLDERS  BE  LIABLE FOR  ANY  CLAIM,  DAMAGES OR  OTHER
 * LIABILITY, WHETHER IN AN ACTION OF  CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE  OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <array>

#include "FileEraser.hpp"
#include "OverwriteMode.hpp"

#include <jni/UniqueUtfChars.hpp>
#include <nullability/NonNull.hpp>
#include <nullability/Nullable.hpp>
#include <time/StopWatch.hpp>

namespace {
    jclass fileExceptionClass = nullptr;
    jclass overwriteModeClass = nullptr;

    jfieldID simpleModeFieldId = nullptr;
    jmethodID nameMethodId = nullptr;
}

namespace firearrow::fs {
    using namespace nullability;
    using namespace time;

    jlong nativeFileEraserCreate(JNIEnv* env, jclass clazz) {
        auto* fileEraser = new FileEraser();
        return reinterpret_cast<jlong>(fileEraser);
    }

    void nativeFileEraserDestroy(JNIEnv* env, jclass clazz, jlong nativeHandle) {
        auto* fileEraser = reinterpret_cast<FileEraser*>(nativeHandle);
        delete fileEraser;
    }

    jlong nativeEraseFile(JNIEnv* rawEnv, jclass clazz, jlong nativeHandle, jstring jvmPath, jobject jvmOverwriteMode) {
        auto* fileEraser = reinterpret_cast<FileEraser*>(nativeHandle);

        if (fileEraser == nullptr) {
            return JNI_ERR;
        }

        auto env = makeNonNull(rawEnv);

        jni::UniqueUtfChars jvmUniquePath(env, jvmPath);
        std::filesystem::path filePath(static_cast<const char*>(jvmUniquePath.get()));

        auto modeName = (jstring) env->CallObjectMethod(jvmOverwriteMode, nameMethodId);
        jni::UniqueUtfChars jvmModeName(env, modeName);

        StopWatch stopWatch;
        stopWatch.start();

        auto overwriteMode = OVERWRITE_MODE.value(jvmModeName.get());

        if (!overwriteMode.has_value()) {
            env->ThrowNew(fileExceptionClass, "Set unknown OverwriteMode");
            return JNI_ERR;
        }

        if (auto result = fileEraser->init(filePath, *overwriteMode); result.hasError()) {
            std::string& message = result.error().message;
            env->ThrowNew(fileExceptionClass, message.c_str());
            return JNI_ERR;
        }

        if (auto result = fileEraser->checkPermission(); result.hasError()) {
            std::string& message = result.error().message;
            env->ThrowNew(fileExceptionClass, message.c_str());
            return JNI_ERR;
        }

        if (auto result = fileEraser->overwriteFile(); result.hasError()) {
            std::string& message = result.error().message;
            env->ThrowNew(fileExceptionClass, message.c_str());
            return JNI_ERR;
        }

        if (auto result = fileEraser->truncateFile(0); result.hasError()) {
            std::string& message = result.error().message;
            env->ThrowNew(fileExceptionClass, message.c_str());
            return JNI_ERR;
        }

        if (auto result = fileEraser->removeFile(); result.hasError()) {
            std::string& message = result.error().message;
            env->ThrowNew(fileExceptionClass, message.c_str());
            return JNI_ERR;
        }

        stopWatch.stop();

        return stopWatch.getDuration();
    }

    jlong nativeEraseFileWithDefaultMode(JNIEnv* rawEnv, jclass clazz, jlong nativeHandle, jstring jvmPath) {
        auto env = makeNonNull(rawEnv);
        jobject simpleModeObject = env->GetStaticObjectField(overwriteModeClass, simpleModeFieldId);

        return nativeEraseFile(env, clazz, nativeHandle, jvmPath, simpleModeObject);
    }

    jlong nativeEraseDirectory(JNIEnv* rawEnv, jclass clazz, jlong nativeHandle, jstring jvmPath, jobject jvmOverwriteMode, jboolean isRecursive) {
        auto env = makeNonNull(rawEnv);
        const auto jvmUniquePath = jni::UniqueUtfChars(env, jvmPath);
        const auto folder = std::filesystem::path(static_cast<const char*>(jvmUniquePath.get()));

        jstring filePath = nullptr;
        StopWatch stopWatch;
        stopWatch.start();

        if (!std::filesystem::exists(folder)) {
            env->ThrowNew(fileExceptionClass, "Directory doesn't exist");
            return JNI_ERR;
        }

        if (!std::filesystem::is_directory(folder)) {
            env->ThrowNew(fileExceptionClass, "Path doesn't directory");
            return JNI_ERR;
        }

        if (isRecursive) {
            for (const auto& item : std::filesystem::recursive_directory_iterator(folder)) {
                filePath = env->NewStringUTF(item.path().c_str());

                if (!std::filesystem::is_directory(item.path())) {
                    if (!nativeEraseFile(env, clazz, nativeHandle, filePath, jvmOverwriteMode)) {
                        return JNI_ERR;
                    }
                }
            }
        } else {
            for (const auto& item : std::filesystem::directory_iterator(folder)) {
                filePath = env->NewStringUTF(item.path().c_str());

                if (!nativeEraseFile(env, clazz, nativeHandle, filePath, jvmOverwriteMode)) {
                    return JNI_ERR;
                }
            }
        }

        stopWatch.stop();

        return stopWatch.getDuration();
    }

    jlong nativeEraseDirectoryWithDefaultMode(JNIEnv* rawEnv, jclass clazz, jlong nativeHandle, jstring jvmPath, jboolean isRecursive) {
        auto env = makeNonNull(rawEnv);
        jobject simpleModeObject = env->GetStaticObjectField(overwriteModeClass, simpleModeFieldId);

        return nativeEraseDirectory(env, clazz, nativeHandle, jvmPath, simpleModeObject, isRecursive);
    }

    constexpr std::array<JNINativeMethod, 6> JNI_METHODS = {{
        {"nativeCreate", "()J", (void*)nativeFileEraserCreate},
        {"nativeDestroy", "(J)V", (void*)nativeFileEraserDestroy},
        {"nativeEraseFile", "(JLjava/lang/String;)J", (void*)nativeEraseFileWithDefaultMode},
        {"nativeEraseFile", "(JLjava/lang/String;Lorg/kl/firearrow/fs/OverwriteMode;)J", (void*)nativeEraseFile},
        {"nativeEraseDirectory", "(JLjava/lang/String;Z)J", (void*)nativeEraseDirectoryWithDefaultMode},
        {"nativeEraseDirectory", "(JLjava/lang/String;Lorg/kl/firearrow/fs/OverwriteMode;Z)J", (void*)nativeEraseDirectory}
    }};
}

jint registerFileManager(JNIEnv* env) {
    using firearrow::fs::JNI_METHODS;

    jclass temporaryClass = env->FindClass("org/kl/firearrow/fs/FileException");
    fileExceptionClass = (jclass) env->NewGlobalRef(temporaryClass);

    if (fileExceptionClass == nullptr) {
        return JNI_ERR;
    }

    temporaryClass = env->FindClass("org/kl/firearrow/fs/OverwriteMode");
    overwriteModeClass = (jclass) env->NewGlobalRef(temporaryClass);

    if (overwriteModeClass == nullptr) {
        return JNI_ERR;
    }

    simpleModeFieldId = env->GetStaticFieldID(overwriteModeClass, "SIMPLE_MODE", "Lorg/kl/firearrow/fs/OverwriteMode;");
    nameMethodId = env->GetMethodID(overwriteModeClass, "name", "()Ljava/lang/String;");

    if (simpleModeFieldId == nullptr || nameMethodId == nullptr) {
        return JNI_ERR;
    }

    jclass fileManagerClass = env->FindClass("org/kl/firearrow/fs/FileManager");

    if (fileManagerClass == nullptr) {
        return JNI_ERR;
    }

    env->DeleteLocalRef(temporaryClass);

    return env->RegisterNatives(fileManagerClass, JNI_METHODS.data(), JNI_METHODS.size());
}

void unregisterFileManager(JNIEnv* env) {
    env->DeleteGlobalRef(overwriteModeClass);
    env->DeleteGlobalRef(fileExceptionClass);
}

